### ISSUE TITLE
Support for QT?

### ISSUE THREAD

**Issue by firelizzard18** [wailsapp/wails#2793]
### Is your feature request related to a problem? Please describe.

I'm curious how much work it would be to update Wails to optionally build against QT and QtWebEngine instead of GTK/WebKit GTK, on the scale of "That could be done with one MR" to "That would take months", and whether it might be feasible for me to contribute this change.

### Describe the solution you'd like

I'd like to build a Wails app for Linux using QT instead of GTK. Partially due to #2792 and partially because I'm using KDE and it would be nice to have the option of using QT.

### Describe alternatives you've considered

N/A

**Comment by leaanthony** [comment]
I don't know the answer off hand but moving away from webkit2gtk would be amazing. The fact that Tauri still uses it suggests it's not simple. I'd fully support this effort though.

**Comment by Psycomentis06** [comment]
Tauri team have been away from any chromium related tech. and this is what one of the creators mentioned. Tauri was built for Linux with FSF guidelines in mind that's why they didn't chose chromium and went for WebKitGtk (for other operating systems users don't care much about FSF so they used what's available for the OS). Tauri team wants to switch from WebKit as well that's why they announced on [Twitter](https://twitter.com/TauriApps/status/1505618001126731781?t=I9IYzsJCURN0TN10MW_L2A&s=19) That they are collaborating with Servo team to make a stable webview provider for all platforms.

PS: I'm using the KDE Falcon browser every now and then which is based on QtWebEngine and I can confirm that i had some problems with some websites that uses CSS to toggle menus/themes ... AKA most Tailwind based UI libraries. So i guess your best choice for now is WebKitGtk until Servo becomes a thing.

**Comment by firelizzard18** [comment]
@Psycomentis06 Beyond non-technical reasons like licensing and FSF/not supporting the monstrosity that is Google, it's not clear to me that either QtWebEngine or WebKit GTK is the better choice. WebKit GTK has its own problems, especially #2792. It seems to me that Wails could easily allow the developer to select between WebKit GTK and QtWebEngine via tags (if someone added QT support).

**Comment by Psycomentis06** [comment]
I agree with you It's good to give developers more options but it's more work. I hope if there is someone interested that add this feature.

**Comment by Blquinn** [comment]
Is there any reason to not just replace gtk with qt? Aside from the ethical browser wars debate. Seems like most users will get a much better experience with a Qt and I'm pretty sure [more people are actually using KDE than gnome anyway](https://www.gamingonlinux.com/users/statistics/), they'll already have Qt installed. (I suspect most gnome users have it installed too)

The tauri team is closing [related issues](https://github.com/tauri-apps/tauri/issues/3988) and quite literally crossing their fingers that webkit performance will improve. No offense to their efforts intended, but it just doesn't seem like they believe there is a real issue.

Also like others have said there's actual rendering and usability issues as well.

**Comment by firelizzard18** [comment]
> I don't know the answer off hand but moving away from webkit2gtk would be amazing. The fact that Tauri still uses it suggests it's not simple. I'd fully support this effort though.

@leaanthony (comment above) clearly supports it, at least in theory. IMO supporting both and letting the developer decide would be best. But regardless, someone has to write that code.

**Comment by leaanthony** [comment]
There is also https://github.com/SerenityOS/serenity/tree/master/Ladybird

**Comment by Blquinn** [comment]
I made a quick proof of concept for a minimal go qt binding here https://github.com/Blquinn/go-qt-binding

I would be interested in providing a qt binding.. I looked through the current linux code and it doesn't seem too difficult. I just can't figure out how to actually build the project and get something running with the locally built wails code. Is there any info on that? Can someone point me in the right direction?

**Comment by Blquinn** [comment]
I was actually able to get a test app running (sort of) so I should be able to make some progress. I've also joined the discord though, thanks!

Side note -- the gtk4 version of webkitgtk-6.0 [has been considered stable for a few months now](https://blogs.gnome.org/mcatanzaro/2023/03/21/webkitgtk-api-for-gtk-4-is-now-stable/) maybe it would be worth upgrading that, or easier than a qt port. From what I understand the performance of the gtk4 version is greatly improved because they're both hardware rendered and gtk is now able to share a framebuffer with webkit... or something like that I'm not super well versed on that.

**Comment by leaanthony** [comment]
Would be happy to upgrade to v4.
Is it installed by default on most distros? We need to think about graceful fallback.

**Comment by firelizzard18** [comment]
If the UI abstraction layer is organized as or could be refactored into 'drivers' for different operating systems and/or libraries, it should be feasible to create drivers for GTK3, GTK4, and Qt, and automatically choose one and/or let the user decide.

**Comment by Blquinn** [comment]
I spent some time implementing a proof of concept for the qt frontend here https://github.com/wailsapp/wails/pull/3024

For now, it just opens a window with example.org, but it handles most of the application life cycle and threading stuff pretty well already.

Let me know what you all think!

> Would be happy to upgrade to v4.
> Is it installed by default on most distros? We need to think about graceful fallback.

I'd say gtk4 is installed on most distros, I'm not sure about the new webkit. I don't think it's made it everywhere yet, definitely not like debian etc. I think you'd likely want to have it as a build tag type situation and support both for a time.

**Comment by leaanthony** [comment]
I actually think this is a packaging issue. If we pakcage as AppImage/Flatpak/snap then that guarantees the correct version being available.

**Comment by leaanthony** [comment]
> If the UI abstraction layer is organized as or could be refactored into 'drivers' for different operating systems and/or libraries, it should be feasible to create drivers for GTK3, GTK4, and Qt, and automatically choose one and/or let the user decide.

Build tags should be sufficient I think.

**Comment by PylotLight** [comment]
> I actually think this is a packaging issue. If we pakcage as AppImage/Flatpak/snap then that guarantees the correct version being available.

Vote 1 Flatpak

**Comment by firelizzard18** [comment]
> I actually think this is a packaging issue. If we pakcage as AppImage/Flatpak/snap then that guarantees the correct version being available.

Being able to distribute my app as a single executable is a huge win for me. App packages are useful but I don't want to be forced to use them.

**Comment by leaanthony** [comment]
For Linux, you have no choice. You aren't currently going to have a single binary with all the dependencies inside and a desktop icon to launch it. The app bundles are trying to solve that.

**Comment by firelizzard18** [comment]
I mean, as a single binary that links to shared libraries on the user's system. I would be quite disappointed if Wails moved to a packaged-app-only model.

**Comment by leaanthony** [comment]
If you can come up with a way to make that work, we'd be all ears. What if the host doesn't have QT installed, or the right version of QT?

**Comment by leaanthony** [comment]
Nobody has suggested we're doing that.

**Comment by rcalixte** [comment]
Hopefully relevant to moving this forward, there are [Qt bindings for Go](https://github.com/mappu/miqt) that support both Qt5 and Qt6, including [Qt WebEngine](https://raw.githubusercontent.com/mappu/miqt/refs/heads/master/examples/libraries/qt6-webengine/screenshot.png). (Qt5 will likely be EOL some time in 2025 so Qt6 should be the point of focus if at all.) These bindings have reached decent maturity with some field testing but could always use more. The author has also been very responsive and helpful. The licenses for Wails and the bindings are also compatible.

> If you can come up with a way to make that work, we'd be all ears. What if the host doesn't have QT installed, or the right version of QT?

Isn't this the same thing with GTK dependencies? When the software is packaged for a distribution, dependencies should be properly specified so that the software can run as expected.

> I don't know the answer off hand but moving away from webkit2gtk would be amazing. The fact that Tauri still uses it suggests it's not simple. I'd fully support this effort though.

It's certainly easier for Go than for Rust. 😉 



> Is it installed by default on most distros?

[Repology](https://repology.org/) is a fantastic resource for determining which version of what software is available for each distro. Targeting a base minimum across the collective latest stable offerings from the major distros should suffice for them and their derivatives. (Arch, Debian, Fedora, and their downstream projects comprise the majority.)




> If you ping me on discord, we will absolutely help you with this 👍

I'm not a fan of the platform but I can join to see if it would help to move this topic forward. I think it would be a huge step forward for this project to be able to at least offer an alternative toolkit (without too much maintenance burden of course).

**Comment by leaanthony** [comment]
Thanks for reaching out\! 👋

This appears to be a question rather than a bug report or feature request. We use [GitHub Discussions](https://github.com/wailsapp/wails/discussions) for Q&A to keep issues focused on actionable items.

**Please feel free to:**
- 💬 Ask in [Q&A Discussions](https://github.com/wailsapp/wails/discussions/categories/q-a)
- 🎮 Join our [Discord](https://discord.gg/wails) for real-time help
- 📚 Check our [documentation](https://wails.io/docs)

Closing this issue, but the question is still valid - just better suited for Discussions\!
